src/werkzeug/datastructures.py

```diff
@@ -469,6 +469,10 @@ def setlistdefault(self, key, default_list=None):
             default_list = dict.__getitem__(self, key)
         return default_list
 
+    def keys(self):
+        """Return an iterator of keys."""
+        yield from dict.keys(self)
```

Comment:
Can you clarify why this is needed? This seems worse to me.

Current behavior on latest master:
```python
In [1]: from werkzeug.datastructures import MultiDict

In [2]: d = MultiDict([('a', 'b'), ('a', 'c')])

In [3]: d.keys()  # resolves to dict.keys(d), which returns a dict_keys object 👌
Out[3]: dict_keys(['a'])
```

Behavior after this change:
```python
In [1]: from werkzeug.datastructures import MultiDict

In [2]: d = MultiDict([('a', 'b'), ('a', 'c')])

In [3]: d.keys()  # why is this better?
Out[3]: <generator object MultiDict.keys at 0x107102120>
```

Reply:
Werkzeug 1.0.1,
```python
>>> from werkzeug.datastructures import MultiDict
>>> d = MultiDict([('a', 'b'), ('a', 'c')])
>>> type(d.keys)
<class 'method'>
>>> d.keys()
<dict_keyiterator object at 0x110288ae0>
>>> {**d}
{'a': 'b'}
```

Werkzeug 2.0.0rc1
```python
>>> from werkzeug.datastructures import MultiDict
>>> d = MultiDict([('a', 'b'), ('a', 'c')])
>>> type(d.keys)
<class 'builtin_function_or_method'>
>>> d.keys()
dict_keys(['a'])
>>> {**d}
{'a': ['b', 'c']}
```

The unpacking behavior has changed between the versions, I think using the builtin keys method triggers an optimised path hence the difference. As 1.0.1 defined its own `keys` method (returning a generator), I think 2.0.0 should also do so. (This reverts part of the changes made to drop Python2 support).

Reply:
That is fascinating. Great catch!

1. This exactly the kind of unintended behavior change from refactoring that unit tests are great at catching. Let's add a unit test for the expected result of `{**m}` when `m` is a `MultiDict`, to prevent future regressions.

1. Do you know the exact code path that Python takes through a mapping when `{**m}` is called on it? Would be great to better understand how this is happening, given that `list(m)`, `list(m.keys())`, and `list(m.items())` are all producing the same results before and after your changes.

Reply:
1. Yep, added. (Oddly I had it locally but missed pushing).
2. I'm not sure, can't quite follow [this](https://github.com/python/cpython/commit/025e9ebd0a0a19f50ca83af6ada0ac65be1fa2a1).

Reply:
odd, this was fixed earlier....

Reply:
It looks like `__iter__` is the method we actually want to override? With this patch applied locally:
```diff
+    def __iter__(self):
+        yield from dict.keys(self)
```
I'm getting the expected result:
```python
In [1]: from werkzeug.datastructures import MultiDict

In [2]: m = MultiDict([('a', 'b'), ('a', 'c')])

In [3]: {**m}
Out[3]: {'a': 'b'}
```

Reply:
Thanks, fixed. Looks like `__iter__` was mapped to `keys` [previously](https://github.com/pallets/werkzeug/blob/1.0.x/src/werkzeug/datastructures.py#L569). I must have messed up my environment.